include/EVT/io/SPI.hpp

```diff
@@ -0,0 +1,143 @@
+#ifndef _EVT_SPI_
+#define _EVT_SPI_
+
+#include "GPIO.hpp"
+#include <stdint.h>
+
+#define SPI_MSB_FIRST 1
+#define SPI_LSB_FIRST 0
+
+#define SPI_MODE0 0x01
+#define SPI_MODE1 0x02
+#define SPI_MODE2 0x04
+#define SPI_MODE3 0x08
+
+#define SPI_SPEED_4MHZ 4000000
+#define SPI_SPEED_2MHZ 2000000
+#define SPI_SPEED_1MHZ 1000000
+#define SPI_SPEED_500KHZ 500000
+#define SPI_SPEED_250KHZ 250000
+#define SPI_SPEED_125KHZ 125000
+#define SPI_SPEED_62KHZ 62500
+#define SPI_SPEED_31KHZ 31250
+
+namespace EVT::core::IO {
+// Forward declarations:
+// The different pins are hardware specific. Forward declaration to allow
+// at compilation time the decision of which pins should be used.
+enum class Pin;
+
+class SPI {
+public:
+    /**
+         * Make an instance of an I2C interface that will use the given pins
+         * for clock and data lines.
+         *
+         * @param sckPin[in] The SPI clock pin
+         * @param mosiPin[in] The mosi pin (data out)
+         * @param misoPin[in] The miso pin (data in)[optional]
+         */
+    SPI(GPIO** CSPins, uint8_t pinLength, Pin sckPin, Pin mosiPin, Pin misoPin);
```

Comment:
Somethings to note.

Having the CSPins passed in and set via this manner will require the array of device pins to remain valid memory for the life time of the SPI device. An example where this might not be the case is if the SPI device is created within a function with the device pins stored as a local variable.

In general, you'll see in embedded development, the parameter will be an array of device pins, and those pins will then be copied into an array of device pins that is stored within the SPI device. That way, you know the device pins will live as long as the SPI interface. That does mean we will also have an upper bound on how many devices will be selected. But below is some example psuedo code

```
SPI(GPIO CSPins[], uint8_t numPins) {
    for(int i = 0; i < numPins; i++) {
        this->pins[i] = CSpins[i];
    }
}
```
Where this->pins would be an instance variable with a maximum size.

Another thing that should be considered is instead of having a list of pins as an input. The SPI device could take a `Mux` object as an input. The idea being that the `Mux` could be implemented in different ways (via inheritance). That way if say the chip select is actually done through a GPIO expander (common) or demux (very common), the SPI device is immune to those changes. That being said, that can be a feature/modification added later.

Also this was a huge wall of text, so feel free to ask for clarification.

Reply:
Sounds good. Since we aren't using much SPI right now, would 5 be a good maximum size for now or should we start larger?

Reply:
Yeah I think 5 is reasonable, you can have it in a `constexpr` so it's easy to update later

Reply:
cool